{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exercise session nº 1 \n",
    "---\n",
    "# Somites Patterning : Autoinhibition with Transcriptional Delay \n",
    "\n",
    "This subject is extracted from : \n",
    "> Julian Lewis, *Autoinhibition with Transcriptional Delay: A Simple Mechanism for the Zebrafish Somitogenesis Oscillator*, Current Biology, 2003.\n",
    "> https://doi.org/10.1016/S0960-9822(03)00534-7\n",
    "\n",
    "Somites are body segments containing the same internal structures. They can be observed at adult age in invertebrates such as earthworms but can also be observed during the embryonic stages of vertebrates. This patterning of these somites is done by an autohibitory mechanism, where two linked genes, her1 and her7, codes for proteins that inhibits themselves, giving rise to oscillations in their expressions. The Delta-Notch signaling pathway provides synchronization of adjacent cells. \n",
    "\n",
    "\n",
    "![alt text](Images/somites.png \"Title\")\n",
    "\n",
    "During these session, we will show how a simple mechanical model can express these oscillations, and show that such a system can prove particularly robust, even to noise, and give the same phenotype with very different biological parameters.\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## I - The single-cell single-gene oscillator\n",
    "\n",
    "The simplest possibility one might envisage would be a feedback loop in which her1 or her7 protein directly binds to the regulatory DNA of its own gene to inhibit transcription. The delay T is linked to the production of biological material.\n",
    "\n",
    "![alt text](Images/Oscillations.png \"subtitle\")\n",
    "\n",
    "We note : $m(t)$ the quantity of mRNA of her1 at time t, and $p(t)$ the quantity of protein her1 at time t. \n",
    "\n",
    "In our situation, however, there are two delays : $T_m$, the time necessary to produce mature mRNAs, and $T_p$, the time needed to translate these mRNAs into proteins. \n",
    "\n",
    "![alt text](Images/Biological_model.png \"subtitle\")\n",
    "\n",
    "We note $b$ and $c$ the decay rates (inverse lifetimes) of $p$ and $m$. \n",
    "\n",
    "We thus obtain the following system of coupled differential equations : \n",
    "\n",
    "$\\dfrac{dp(t)}{dt} = am(t-T_p) - bp(t)$\n",
    "\n",
    "$\\dfrac{dm(t)}{dt} = f(p(t-T_m))-cm(t)$\n",
    "\n",
    "We will chose $f(p)=\\dfrac{k}{(\\frac{p}{p_{crit}})^2+1}$, where k is the number of mRNA molecules synthetised without inhibition, and $p_{crit}$ is the concentration of protein which gives half-maximal inhibition. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__These equations are numerically stable, they can be discretized using a forward euler-scheme__\n",
    "\n",
    "#### **Question 1 :**\n",
    "> Discretize these differential equations with a forward euler-scheme\n",
    "\n",
    "\n",
    "#### **Question 2 :**\n",
    "> Integrate these equations for t<tfinal. Try with a = 4.5, a = 1.5, a = 0.5. Which behaviours are induced by this change of parameter ? \n",
    "*Hint : Make use of a 1d linear interpolation to compute $m(t-T_p)$ and $p(t-T_m)$ *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The following values are choosen from the mathematical analysis of the paper and is adapted for biological values of her7\n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.interpolate import interp1d\n",
    "\n",
    "tfinal = 500\n",
    "dt = 0.2\n",
    "Tm = 7.1 \n",
    "Tp = 1.7\n",
    "\n",
    "pcrit = 40\n",
    "k = 33\n",
    "f = lambda x : k / (1 + (x**2)/(pcrit**2))\n",
    "\n",
    "a_values = [4.5,1.5,0.5]\n",
    "a = a_values[0]\n",
    "b=0.23\n",
    "c=0.23"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nfinal = np.ceil(tfinal/dt)\n",
    "npoints = int(nfinal)+1\n",
    "timepoints = np.linspace(0,tfinal,npoints)\n",
    "\n",
    "P = np.zeros(npoints)\n",
    "M = np.zeros(npoints)\n",
    "\n",
    "for nt,t in enumerate(timepoints):\n",
    "        if nt==0 : \n",
    "            continue\n",
    "        if t< Tp \n",
    "        ...\n",
    "        else : \n",
    "            \n",
    "        if t<Tm\n",
    "        ...\n",
    "        else :\n",
    "        ...\n",
    "        \n",
    "        ## Update P and M Values\n",
    "            \n",
    "        P[nt]=...\n",
    "        M[nt]=...\n",
    "        ...\n",
    "\n",
    "\n",
    "## Plot the results : \n",
    "plt.plot(timepoints,P,'tab:red')\n",
    "plt.plot(timepoints,M,'tab:blue')\n",
    "plt.xlim(0,500)\n",
    "plt.ylim(0,350)\n",
    "plt.title(\"a:\"+str(a)+\"  b:0.23   c:0.23  k:33.  delaym:7.1.  delayp:1.7\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## II - Noisy Control of Gene expression\n",
    "\n",
    "In any transcriptional control system, random fluctuations will arise from random variation, or noise, in the regulation of gene expression [31]. The binding and dissociation of a gene regulatory protein to and from its site on DNA are stochastic processes, each described by a certain probability of occurrence per unit time, so that we should expect her1 and her7 to flicker between “off” (repressor bound) and “on” (repressor dissociated) states. In writing down our deterministic equations that describe the gene regulation in terms of a smooth dependence of transcription rate on protein concentration, we are in effect assuming that we can replace the underlying on/off flickering behavior by an instantaneous time average, in which the rate of transcription is proportional to the fraction of time the gene is currently spending with repressor protein bound to it. The slower the average flicker rate, the more the real system will depart from the idealized deterministic behavior\n",
    "\n",
    "We now add to our model the possibility for the protein to bind and unbind randomly from its DNA site. \n",
    "These binding rate is noted $k_{on}$ and the unbinding rate is noted $k_{off}$\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A - Haploid Expression\n",
    "\n",
    "We define X[t] = number of gene copies that are free of repression at time t. It can be of value 0, 1 for and haploid cell.\n",
    "\n",
    "Lets consider only one copy of the gene. \n",
    "\n",
    "We define $D(t)$ the probability that the gene is free of repression, i.e the probability that X(t) =1. \n",
    "\n",
    "### Question 3 : \n",
    "> Show that $\\dfrac{dD}{dt} = k_{off}(1-D) - k_{on}D$\n",
    "\n",
    "\n",
    "\n",
    "In the stochastic system, we assume that transcripts are initiated at zero rate when repressor is bound and at maximal rate when it is not bound. Then to make the correct correspondence between the noisy system and the deterministic system, we must have : \n",
    "\n",
    "$kD = f(p)=\\dfrac{k}{(\\frac{p}{p_{crit}})^2+1}$\n",
    "\n",
    "Thus we need to have $k \\dfrac{k_{off}}{k_{on}+k_{off}} =\\dfrac{k}{(\\frac{p}{p_{crit}})^2+1}$\n",
    "\n",
    "Which gives the condition $k_{on} = k_{off}(\\frac{p}{p_{crit}})^2.$\n",
    "\n",
    "---\n",
    "\n",
    "The time series $X(t)$ is a markov chain that is defined by the following transition probabilities (We will use the following notations : pij : probability that the state at time t+dt will be i (i.e. that X[t+dt]=i ) if the state at time t is j (i.e. that X[t]=j)): \n",
    "\n",
    "$\\begin{align}\n",
    "p_{00} & = \\mathbb{P}(X(t+dt)=0|X(t)=0) = 1 - D(t+dt)_{|D(t) = 0} \\newline\n",
    "p_{10} & = \\mathbb{P}(X(t+dt)=1|X(t)=0) = D(t+dt)_{|D(t) = 0} \\newline\n",
    "p_{01} & = \\mathbb{P}(X(t+dt)=0|X(t)=1) = 1 - D(t+dt)_{|D(t) = 1} \\newline\n",
    "p_{11} & = \\mathbb{P}(X(t+dt)=1|X(t)=1) = D(t+dt)_{|D(t) = 1} \\newline\n",
    "\\end{align}$\n",
    "\n",
    "If we write $u = \\dfrac{k_{off}}{k_{on}+k_{off}}$ and $v = k_{on} + k_{off}$, we can rewrite, at first order : \n",
    "\n",
    "$D(t+dt) = u(1-e^{-vdt}) + D(t)e^{-vdt}$, thus :\n",
    "\n",
    "$\\begin{align}\n",
    "p_{00} & = 1 - u(1-e^{-vdt}) \\newline\n",
    "p_{10} & = u(1-e^{-vdt}) \\newline\n",
    "p_{01} & = (1 - u)(1-e^{-vdt})\\newline\n",
    "p_{11} & = u(1-e^{-vdt}) + e^{-vdt} \\newline\n",
    "\\end{align}$\n",
    "\n",
    "We can simulate the stochastic process by assigning values to X at successive time points according to the above transition probability rule, and the stochastic system will then tend toward the behaviour of the deterministic continuum system in the limit of large kon and koff , where the rate of stochastic fluctuation (the rate of flickering) is very rapid in comparison with other variables such as the molecular lifetimes.\n",
    "\n",
    "$p(t+dt) = am(t-T_p)dt + p(t)(1-bdt)$\n",
    "\n",
    "$m(t+dt) = k_1(x(t-T_m))dt + m(t)(1-cdt)$\n",
    "\n",
    "\n",
    "\n",
    "#### **Question 4 :**\n",
    "> Integrate these equations for t<tfinal. Try with a = 4.5, a = 0.45, a = 0.225, a = 0.045. Which behaviours are induced by this change of parameter ? Compare with the one without noise. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "minute = 1\n",
    "tfinal = 1000\n",
    "dt = 0.1\n",
    "Tm = 12\n",
    "Tp = 2.8\n",
    "pcrit = 40\n",
    "k = 33\n",
    "k1 = k/2 # k1 is the maximal rate of her1 mRNA synthesis per gene copy, i.e. number of transcripts initiated per unit time per gene copy when inhibitory complex is dissociated from the promote\n",
    "koff = 1 #  off-rate for dissociation of inhibitory complex from promoter of her1 \n",
    "kon = lambda p : koff* (p/pcrit)**2\n",
    "nfinal = np.ceil(tfinal/dt)\n",
    "npoints = int(nfinal)+1\n",
    "f = lambda x : k / (1 + (x**2)/(pcrit**2))\n",
    "\n",
    "def p00(u,v,dt) : \n",
    "    return(1 - u*(1-np.exp(-v*dt)))\n",
    "def p10(u,v,dt) : \n",
    "    return(u*(1-np.exp(-v*dt)))\n",
    "def p01(u,v,dt) : \n",
    "    return((1-u)*(1-np.exp(-v*dt)))\n",
    "def p11(u,v,dt) : \n",
    "    return(u*(1-np.exp(-v*dt))+np.exp(-v*dt))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import interpolate\n",
    "timepoints = np.linspace(0,tfinal,npoints)\n",
    "\n",
    "a = 4.5\n",
    "\n",
    "P = np.zeros(npoints)\n",
    "M = np.zeros(npoints)\n",
    "X = np.zeros(npoints)\n",
    "for nt, t in enumerate(timepoints): \n",
    "    if t==0 : \n",
    "        continue\n",
    "\n",
    "    if Tp < t : \n",
    "        ...\n",
    "    else : \n",
    "        ...\n",
    "\n",
    "    if Tm <t : \n",
    "        ...\n",
    "    else : \n",
    "        ...\n",
    "\n",
    "    ## Update P and M Values\n",
    "    \n",
    "    P[nt] = ...\n",
    "    M[nt] = ...\n",
    "\n",
    "    v = kon(P[nt-1]) + koff\n",
    "    u = koff/v\n",
    "\n",
    "\n",
    "    ## Update X value\n",
    "    if X[nt-1] == 0 : \n",
    "        ...\n",
    "    if X[nt-1] == 1 : \n",
    "        ...\n",
    "\n",
    "\n",
    "\n",
    "plt.plot(timepoints,P/10,'tab:red')\n",
    "plt.plot(timepoints,M,'tab:blue')\n",
    "plt.xlim(0,1000)\n",
    "plt.ylim(0,140)\n",
    "plt.title(\"a:\"+str(a)+\"  b:0.23   c:0.23  k:33.  delaym:7.1.  delayp:1.7\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### B - Diploid Expression\n",
    "We define XX[t] = number of gene copies that are free of repression at time t. It can be of value 0, 1, 2 for a diploid cell.\n",
    "The time series $X(t)$ is a markov chain that is defined by transition probabilities (We will use the following notations : ppij : probability that the state at time t+dt will be i (i.e. that XX[t+dt]=i ) if the state at time t is j (i.e. that XX[t]=j)). They can be expressed from the haploid case\n",
    "\n",
    "### Question 5 : \n",
    "> Express the diploid transition probabilities $ppij_{i,j \\in (0,1,2)}$ from the haploid ones $pij_{i,j \\in (0,1)}$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Question 6 :**\n",
    "> Integrate these equations for t<tfinal. Try with a = 4.5, a = 0.45, a = 0.225, a = 0.045. Which behaviours are induced by this change of parameter ? How does the noise affects the presence of oscillations ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pp00(u,v,dt):\n",
    "    return(p00(u,v,dt)**2)\n",
    "def pp10(u,v,dt):\n",
    "    return(2*p10(u,v,dt)**p00(u,v,dt))\n",
    "def pp20(u,v,dt):\n",
    "    return(p10(u,v,dt)**2)\n",
    "def pp01(u,v,dt):\n",
    "    return(p01(u,v,dt)*p00(u,v,dt))\n",
    "def pp11(u,v,dt):\n",
    "    return(p11(u,v,dt)*p00(u,v,dt) + p01(u,v,dt)*p10(u,v,dt))\n",
    "def pp21(u,v,dt):\n",
    "    return(p10(u,v,dt)*p11(u,v,dt))\n",
    "def pp02(u,v,dt):\n",
    "    return(p01(u,v,dt)**2)\n",
    "def pp12(u,v,dt):\n",
    "    return(2*p01(u,v,dt)*p11(u,v,dt))\n",
    "def pp22(u,v,dt):\n",
    "    return(p11(u,v,dt)**2)\n",
    "\n",
    "\n",
    "from scipy import interpolate\n",
    "timepoints = np.linspace(0,tfinal,npoints)\n",
    "\n",
    "a=4.5\n",
    "P = np.zeros(npoints)\n",
    "M = np.zeros(npoints)\n",
    "XX = np.zeros(npoints)\n",
    "for nt, t in enumerate(timepoints): \n",
    "    if t==0 : \n",
    "        continue\n",
    "\n",
    "    if Tp < t : \n",
    "        ...\n",
    "    else : \n",
    "        ...\n",
    "\n",
    "    if Tm <t : \n",
    "        ...\n",
    "    else : \n",
    "        ...\n",
    "        \n",
    "    # Update P and M values\n",
    "\n",
    "    P[nt] = ...\n",
    "    M[nt] = ...\n",
    "\n",
    "    v = kon(P[nt-1]) + koff\n",
    "    u = koff/v\n",
    "\n",
    "\n",
    "    # Update XX values\n",
    "    \n",
    "    if XX[nt-1] == 0 : \n",
    "        ...\n",
    "    if XX[nt-1] == 1 : \n",
    "        ...\n",
    "    if XX[nt-1] == 2 : \n",
    "        ...\n",
    "\n",
    "\n",
    "plt.plot(timepoints,P/10,'tab:red')\n",
    "plt.plot(timepoints,M,'tab:blue')\n",
    "plt.xlim(0,1000)\n",
    "plt.ylim(0,140)\n",
    "plt.title(\"a:\"+str(a)+\"  b:0.23   c:0.23  k:33.  delaym:7.1.  delayp:1.7\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## III - Synchronisation of the oscillations through Delta-Notch pathway : The case of two coupled cells \n",
    "\n",
    "\n",
    "The author hypothesize that deltaC has a expression pattern as her1 and her7, as well as a similar lifetime, thus its production will oscillate as well. Delta ligands binds to the Notch receptor of neighboring cells (see figure), which activates a signaling pathway that stimulates the expression of her1/her7. We will see that the coupling of two cells with a slightly different oscillating period can be synchronized with this coupling. \n",
    "\n",
    "![alt text](Images/Synchronisation_tiny.png \"subtitle\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In each of the two cells, the concentration of the proteins are governed by the following equations :\n",
    "\n",
    "$\n",
    "\\begin{align}\n",
    "\\frac{dp_{her1}}{dt} & = a_{her1} m_{her1}(t-T_{pher1}) - b_{her1} p_{her1}(t) \\newline\n",
    "\\frac{dp_{her7}}{dt} & = a_{her7} m_{her7}(t-T_{pher7}) - b_{her7} p_{her7}(t) \\newline\n",
    "\\frac{dp_{delta}}{dt} & = a_{delta} m_{delta}(t-T_{delta}) - b_{delta} p_{delta}(t) \\newline\n",
    "\\frac{dm_{her1}}{dt} & = f_{her1}(p_{her1},p_{her7},\\tilde{p}_{delta})_{|t-T_{mher1}} - c_{her1}m_{her1}(t)\\newline\n",
    "\\frac{dm_{her7}}{dt} & = f_{her7}(p_{her1},p_{her7},\\tilde{p}_{delta})_{|t-T_{mher7}} - c_{her7}m_{her7}(t)\\newline\n",
    "\\frac{dm_{delta}}{dt} & = f_{delta}(p_{her1},p_{her7},\\tilde{p}_{delta})_{|t-T_{mdelta}} - c_{delta}m_{delta}(t) \\newline\n",
    "\\end{align}\n",
    "$\n",
    "\n",
    "Where $\\tilde{}$ indicates the concentration of the neighboring protein, and the functions are defined by : \n",
    "\n",
    "$\n",
    "\\begin{align}\n",
    "f_{her1}(p_{her1},p_{her7},\\tilde{p}_{delta}) &\n",
    "= k_{her1} (r_0 + (1-r_0)\\frac{\\tilde{\\phi}_d}{1+\\tilde{\\phi}_d}\\frac{1}{1+\\phi_{her1}\\phi_{her7}}) \\newline\n",
    "f_{her7}(p_{her1},p_{her7},\\tilde{p}_{delta}) &\n",
    "= k_{her7} (r_0 + (1-r_0)\\frac{\\tilde{\\phi}_d}{1+\\tilde{\\phi}_d}\\frac{1}{1+\\phi_{her1}\\phi_{her7}}) \\newline\n",
    "f_{delta}(p_{her1},p_{her7},\\tilde{p}_{delta}) &\n",
    "= k_{delta} (s_h \\frac{1}{1+\\phi_{her1}\\phi_{her7}}+ (1-s_h)\\frac{\\tilde{\\phi}_d}{1+\\tilde{\\phi}_d}\\frac{1}{1+\\phi_{her1}\\phi_{her7}})\n",
    "\\end{align}\n",
    "$\n",
    "\n",
    "Where $\\phi_x$ is the reduced value of each protein concentration, where the concentrations are divided by their critical values. \n",
    "\n",
    "#### **Question 7 :**\n",
    "> Integrate these equations for t<tfinal. Try to remove the coupling between the two cells. What do you observe ? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#parameters :\n",
    "\n",
    "minute = 1 \n",
    "eps = 0.3\n",
    "dt = eps\n",
    "tfinal = 600 # Duration of the simulation\n",
    "ah1,ah7,ad = [4.5]*3 # Protein synthesis rates per mRNA molecule for Her1, Her7 and Delta, for the two cells\n",
    "bh1,bh7,bd = [0.23]*3 # Protein degradation rates : inverse lifetimes for Her1, Her7 and Delta\n",
    "ch1,ch7,cd = [0.23]*3 # mRNA degradation rates : inverse lifetimes for Her1, Her7 and Delta\n",
    "delaymh1 = 12*np.array([0.95,1.05])\n",
    "delayph1 = 2.8*np.array([1,1])\n",
    "delaymh7 = 7.1*np.array([0.95,1.05])\n",
    "delayph7 = 1.7*np.array([1,1])\n",
    "delaymd = 16*np.array([1,1])\n",
    "delaypd = 20.5*np.array([1,1])\n",
    "\n",
    "kh1 = 33\n",
    "kh7 = 33\n",
    "kd = 33\n",
    "critph = 40 \n",
    "critpd = 1000\n",
    "r0 = 0.01\n",
    "rh = 0 \n",
    "rd = 0\n",
    "rhd = 1- r0 -rh - rd\n",
    "s0 = 0 \n",
    "sh = 1\n",
    "sd = 0 \n",
    "shd = 1 - s0 - sh - sd\n",
    "\n",
    "def fr(x1, x7, y) : \n",
    "    return (r0 + rd*y/(1+y) + rh* 1/(1+x1*x7) + rhd* (y/(1+y))*(1/(1+x1*x7)) )\n",
    "def fs(x1,x7,y):\n",
    "    return (s0 + sd*y/(1+y) + sh* 1/(1+x1*x7)  + shd*(y/(1+y))*(1/1+x1*x7) )\n",
    "\n",
    "def fh1(xh1, xh7, yd):\n",
    "    return kh1 * fr(xh1/critph, xh7/critph, yd/critpd)\n",
    "def fh7(xh1, xh7, yd):\n",
    "    return kh7 * fr(xh1/critph, xh7/critph, yd/critpd)\n",
    "def fd(xh1, xh7, yd):\n",
    "    return kd * fs(xh1/critph, xh7/critph, yd/critpd)\n",
    "\n",
    "npoints = int(tfinal/eps)+1\n",
    "timepoints = np.linspace(0,tfinal,npoints)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "for nt, t in enumerate(timepoints): \n",
    "    ..."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
